### PR TITLE
daemon, cli: add pipeline command 'install'

### PR DESCRIPTION
**Pull request by FeelyChau**
Add new command for pipeline: `install`.
```shell
$ pipcook pipeline install https://raw.githubusercontent.com/alibaba/pipcook/master/test/pipelines/text-bayes-classification.json --verbose
```

### CODE REVIEW

File: .github/workflows/cli.yml [hunk 1]
```diff
@@ -67,6 +67,10 @@ jobs:
       run: |
         ./packages/cli/dist/bin/pipcook plugin install ./packages/plugins/data-collect/csv-data-collect
         ./packages/cli/dist/bin/pipcook plugin install ./packages/plugins/model-define/detectron-fasterrcnn-model-define
+    - name: pipcook plugin install from pipeline config
+      run: |
+        ./packages/cli/dist/bin/pipcook plugin installFromConfig ./test/pipelines/text-bayes-classification.json
+        ./packages/cli/dist/bin/pipcook plugin installFromConfig https://raw.githubusercontent.com/alibaba/pipcook/master/test/pipelines/text-bayes-classification.json --verbose
```
Comment (yorkie) [inline .github/workflows/cli.yml:73]: We need another context to test install from config, their plugins are installed at the above commands.

File: packages/cli/src/bin/pipcook-plugin.ts [hunk 2]
```diff
@@ -45,7 +48,62 @@ async function list(opts: any): Promise<void> {
...
+  }
 
+  const params = {
+    cwd: process.cwd(),
+    config: filename,
+    pyIndex: opts.tuna ? tunaMirrorURI : undefined
+  };
+  if (!opts.verbose) {
+    await get(`${route.plugin}/installFromConfig`, params);
+    spinner.succeed(`install plugins succeeded.`);
```
Comment (yorkie) [inline packages/cli/src/bin/pipcook-plugin.ts:80]: IMO the install command will always be pending.

File: packages/cli/src/bin/pipcook-plugin.ts [hunk 3]
```diff
@@ -45,7 +48,62 @@ async function list(opts: any): Promise<void> {
     console.table(plugins, [ 'name', 'version', 'category', 'datatype' ]);
   }
 }
```
Comment (yorkie) [inline packages/cli/src/bin/pipcook-plugin.ts:50]: A blank line is required.

File: packages/cli/src/bin/pipcook-plugin.ts [hunk 4]
```diff
@@ -45,7 +48,62 @@ async function list(opts: any): Promise<void> {
...
+async function installFromConfig(filename: string, opts: any): Promise<void> {
+  const spinner = ora();
+  if (!filename) {
+    spinner.fail('Please specify the config path');
+    return process.exit(1);
+  }
+  let urlObj = url.parse(filename);
+  // file default if the protocol is null
+  if (urlObj.protocol === null) {
```
Comment (yorkie) [inline packages/cli/src/bin/pipcook-plugin.ts:59]:
```suggestion
  if (urlObj.protocol == null) {
```

File: packages/cli/src/bin/pipcook-plugin.ts [hunk 5]
```diff
@@ -45,7 +48,62 @@ async function list(opts: any): Promise<void> {
...
+    if (!existsSync(filename)) {
+      spinner.fail(`${filename} not exists`);
+      return process.exit(1);
+    } else {
+      filename = url.parse(`file://${filename}`).href;
+    }
+  } else if ([ 'http:', 'https:' ].indexOf(urlObj.protocol) === -1) {
+    spinner.fail(`protocol ${urlObj.protocol} is not supported`);
+    return process.exit(1);
+  }
```
Comment (yorkie) [inline packages/cli/src/bin/pipcook-plugin.ts:71]: The above is duplicated with https://github.com/alibaba/pipcook/blob/master/packages/cli/src/actions/start.ts#L21-L33.

File: packages/cli/src/bin/pipcook-plugin.ts [hunk 6]
```diff
@@ -56,7 +114,12 @@ program
     }
     install(name, opts);
   });
-
+program
+  .command('installFromConfig <pipeline>')
```
Comment (yorkie) [inline packages/cli/src/bin/pipcook-plugin.ts:118]: Hmm, I'm -1 on this command-line style, how about:

- `pipcook pipeline install <URL>` is to tell users, install the plugins from a pipeline with URL.

File: packages/daemon/src/app/controller/job.ts [hunk 7]
```diff
@@ -4,6 +4,8 @@ import { PipelineService } from '../../service/pipeline';
 import { parseConfig } from '../../runner/helper';
 import ServerSentEmitter from '../../utils/emitter';
 import { JobModel } from '../../model/job';
+import { PluginManager } from '../../service/plugin';
+import { constants } from '@pipcook/pipcook-core';
```
Comment (yorkie) [inline packages/daemon/src/app/controller/job.ts:8]: Name resolving should be put behind the relative import.

File: packages/daemon/src/app/controller/job.ts [hunk 8]
```diff
@@ -41,10 +46,21 @@ export class JobController {
   }
 
   private async runJobWithContext(job: JobModel, cwd: string, verbose: boolean, pyIndex: string) {
+    const installPlugins = async () => {
+      for (const i in constants.PLUGINS) {
```
Comment (yorkie) [inline packages/daemon/src/app/controller/job.ts:50]:
```suggestion
      for (const type in constants.PLUGINS) {
```

File: packages/daemon/src/app/controller/job.ts [hunk 9]
```diff
@@ -41,10 +46,21 @@ export class JobController {
...
+        }
+        job[plugin] = await this.pluginManager.fetch(job[plugin]);
+        await this.pluginManager.install(job[plugin], pyIndex);
+      }
+    };
     if (verbose) {
       const sse = new ServerSentEmitter(this.ctx);
-      sse.emit('job created', job);
       try {
+        await installPlugins();
```
Comment (yorkie) [inline packages/daemon/src/app/controller/job.ts:62]: Shall we make pending always for installing something?

File: packages/cli/src/bin/pipcook-plugin.ts [hunk 10]
```diff
@@ -6,6 +6,9 @@ import ora from 'ora';
 import { listen, get } from '../request';
 import { route } from '../router';
 import { tunaMirrorURI } from '../config';
+import * as url from 'url';
+import { existsSync } from 'fs';
```
Comment (yorkie) [inline packages/cli/src/bin/pipcook-plugin.ts:10]: Changed to use `pathExists` from `fs-extra`?

File: packages/cli/src/bin/pipcook-plugin.ts [hunk 11]
```diff
@@ -6,6 +6,9 @@ import ora from 'ora';
...
+import { ChildProcess } from 'child_process';
```
Comment (yorkie) [inline packages/cli/src/bin/pipcook-plugin.ts:11]: The core modules should be moved in front of relative imports.

File: .github/workflows/cli.yml (hunk 1 above)
Comment (FeelyChau) [reply .github/workflows/cli.yml:73]: Done.

File: packages/cli/src/bin/pipcook-plugin.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/cli/src/bin/pipcook-plugin.ts:80]: Done.

File: packages/cli/src/bin/pipcook-plugin.ts (hunk 3 above)
Comment (FeelyChau) [reply packages/cli/src/bin/pipcook-plugin.ts:50]: Done.

File: packages/cli/src/bin/pipcook-plugin.ts (hunk 4 above)
Comment (FeelyChau) [reply packages/cli/src/bin/pipcook-plugin.ts:59]: Done.

File: packages/cli/src/bin/pipcook-plugin.ts (hunk 5 above)
Comment (FeelyChau) [reply packages/cli/src/bin/pipcook-plugin.ts]: Done.

File: packages/daemon/src/app/controller/job.ts (hunk 7 above)
Comment (FeelyChau) [reply packages/daemon/src/app/controller/job.ts]: Done.

File: packages/daemon/src/app/controller/job.ts (hunk 8 above)
Comment (FeelyChau) [reply packages/daemon/src/app/controller/job.ts]: Done.

File: packages/daemon/src/app/controller/job.ts (hunk 9 above)
Comment (FeelyChau) [reply packages/daemon/src/app/controller/job.ts]: Done.

File: packages/cli/src/bin/pipcook-plugin.ts (hunk 11 above)
Comment (FeelyChau) [reply packages/cli/src/bin/pipcook-plugin.ts]: Done.

File: packages/cli/src/bin/pipcook-plugin.ts (hunk 6 above)
Comment (FeelyChau) [reply packages/cli/src/bin/pipcook-plugin.ts]: Done.

**Comment by FeelyChau** [comment]
Fail...again!

**Comment by yorkie** [comment]
> Fail...again!

```
Collecting jieba
  Using cached https://files.pythonhosted.org/packages/c6/cb/18eeb235f833b726522d7ebed54f2278ce28ba9438e3135ab0278d9792a2/jieba-0.42.1.tar.gz
Installing collected packages: jieba
  Running setup.py install for jieba: started
    Running setup.py install for jieba: finished with status 'done'
Successfully installed jieba-0.42.1
Collecting scikit-learn
  Using cached https://files.pythonhosted.org/packages/32/37/51bf33e269be272255ab6a1c172e15b930bcf33316616efc82e0759ca4ff/scikit_learn-0.23.1-cp37-cp37m-macosx_10_9_x86_64.whl
Collecting joblib>=0.11 (from scikit-learn)
  Using cached https://files.pythonhosted.org/packages/b8/a6/d1a816b89aa1e9e96bcb298eb1ee1854f21662ebc6d55ffa3d7b3b50122b/joblib-0.15.1-py3-none-any.whl
Collecting scipy>=0.19.1 (from scikit-learn)
  Using cached https://files.pythonhosted.org/packages/85/7a/ae480be23b768910a9327c33517ced4623ba88dc035f9ce0206657c353a9/scipy-1.4.1-cp37-cp37m-macosx_10_6_intel.whl
Collecting numpy>=1.13.3 (from scikit-learn)
  Downloading https://files.pythonhosted.org/packages/3e/00/0266fefaafb839760d5b25b884375b2ab0f842ebe138ee6c1ef807af44bb/numpy-1.18.5-cp37-cp37m-macosx_10_9_x86_64.whl (15.1MB)
Collecting threadpoolctl>=2.0.0 (from scikit-learn)
  Using cached https://files.pythonhosted.org/packages/f7/12/ec3f2e203afa394a149911729357aa48affc59c20e2c1c8297a60f33f133/threadpoolctl-2.1.0-py3-none-any.whl
Installing collected packages: joblib, numpy, scipy, threadpoolctl, scikit-learn
    ✗ should install the package with conda packages
      - Error: Timeout - Async function did not complete within 60000ms (custom timeout)
```

**Pull request merged**